### PR TITLE
Handle binding:vnic_type as optional (#55526)

### LINKED ISSUE #55524: os_port exects binding:vnic_type
##### SUMMARY
Ansible fails on an update of an OpenStack port if the server does not report the attribute
`binding:vnic_type` of a port.
Ansible should not assume that the server reports this attribute because this attribute is part of the optional port binding extension of the OpenStack API.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
os_port

##### ANSIBLE VERSION
```paste below
ansible 2.9.0.dev0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible-2.9.0.dev0-py2.7.egg/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15 (default, Oct 15 2018, 15:24:06) [GCC 8.1.1 20180712 (Red Hat 8.1.1-5)]
```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT
fedora 28

##### STEPS TO REPRODUCE

Apply the following playbook two times

```yaml
---
- name: Create network net10
  os_network:
    cloud: "{{ cloud_name }}"
    state: present
    name: net10
  register: net10

- name: Create port net10_port1
  os_port:
    cloud: "{{ cloud_name }}"
    state: present
    name: net10_port1
    network: net10
    fixed_ips:
      - ip_address: 10.10.0.11
```

##### EXPECTED RESULTS

Second run of the playbook succeeds without update.

##### ACTUAL RESULTS

```paste below
al: [localhost]: FAILED! => {"changed": false, "module_stderr": "No handlers could be found for logger \"keystoneauth.identity.generic.base\"\nTraceback (most recent call last):\n  File \"/home/user/.ansible/tmp/ansible-tmp-1555597535.79-177140408209916/AnsiballZ_os_port.py\", line 114, in <module>\n    _ansiballz_main()\n  File \"/home/user/.ansible/tmp/ansible-tmp-1555597535.79-177140408209916/AnsiballZ_os_port.py\", line 106, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/home/user/.ansible/tmp/ansible-tmp-1555597535.79-177140408209916/AnsiballZ_os_port.py\", line 49, in invoke_module\n    imp.load_module('__main__', mod, module, MOD_DESC)\n  File \"/tmp/ansible_os_port_payload_wiDsZ4/__main__.py\", line 410, in <module>\n  File \"/tmp/ansible_os_port_payload_wiDsZ4/__main__.py\", line 393, in main\n  File \"/tmp/ansible_os_port_payload_wiDsZ4/__main__.py\", line 236, in _needs_update\nKeyError: 'binding:vnic_type'\n", "module_stdout": "", "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", "rc": 1}

```

### LINKED ISSUE #55525: os_port sends binding:vnic_type=normal without requested
##### SUMMARY

On creating a new OpenStack port, os_port sends `binding:vnic_type="normal"` without requested.
##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

os_port
##### ANSIBLE VERSION

```paste below
ansible 2.9.0.dev0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/user/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/site-packages/ansible-2.9.0.dev0-py2.7.egg/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.15 (default, Oct 15 2018, 15:24:06) [GCC 8.1.1 20180712 (Red Hat 8.1.1-5)]

```

##### CONFIGURATION

```paste below

```

##### OS / ENVIRONMENT

fedora 28

##### STEPS TO REPRODUCE

Apply the following playbook

```yaml
---
- name: Create network net10
  os_network:
    cloud: "{{ cloud_name }}"
    state: present
    name: net10
  register: net10

- name: Create port net10_port1
  os_port:
    cloud: "{{ cloud_name }}"
    state: present
    name: net10_port1
    network: net10
    fixed_ips:
      - ip_address: 10.10.0.11
```

##### EXPECTED RESULTS

The request to create the port is sent without the attribute `binding:vnic_type` is set.

##### ACTUAL RESULTS

The request to create the port is sent with `binding:vnic_type=normal`

```paste below

```

### PR DESCRIPTION
**Pull request by dominikholler**
According to the OpenStack Networking API
the attribute binding:vnic_type of a port is optional.
This change enables the os_port module to handle
binding:vnic_type as optional.

(cherry picked from commit bc50a52ee23272fc70988139bc9a0d4b287f2cb3)

##### SUMMARY

This is the backport of PR #55526 to ansible-2.8 .

Fixes #55524
Fixes #55525

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

os_port

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Review by cloudnull — APPROVED** [review]

**Comment by cloudnull** [comment]
shipit

File: lib/ansible/modules/cloud/openstack/os_port.py
Comment (abadger) [inline lib/ansible/modules/cloud/openstack/os_port.py:360]: Couple notes:  You probably shouldn't pop from module.params.... I usually consider that to be a "constant" value.  It's the record of what the user specified in the playbook for the module to look at but not to change.

If you do want to pop it, you don't need to do it as two steps.  This should be equivalent to what the present code does:
``` python
        module.params['binding:vnic_type'] = module.params.pop('vnic_type', None)
```

**Comment by abadger** [comment]
This looks like it will be a change in behaviour for the module.... Playbooks which did not specify this parameter before would have had it set to "normal".  Once this change is made, the parameter will not be set instead.

As such, I don't know if it should be backported.  And we probably should have a porting guide entry for it in the 2.9 release as well.

File: lib/ansible/modules/cloud/openstack/os_port.py
Comment (dominikholler) [reply lib/ansible/modules/cloud/openstack/os_port.py:360]: I tried to keep the code change as minimal as possible since this just a bugfix.
Additionally, I am not sure, if it is a good idea to modify a backport. Maybe I can create a code change in the devel branch and add this to this PR?

**Review by dominikholler — COMMENTED** [review]
> This looks like it will be a change in behaviour for the module.... Playbooks which did not specify this parameter before would have had it set to "normal". Once this change is made, the parameter will not be set instead.

> As such, I don't know if it should be backported. And we probably should have a porting guide entry for it in the 2.9 release as well.

Ansible 2.8 is now in the unpleasant state, that it was released
with wrong behavior (issues #55524 and #55525).

Now it is up to the ansible 2.8 maintainers to decide, if they expect
that the implementations of OpenStack Networking API adopt the new
behavior of ansible 2.8, or ansible 2.8 changes its behavior to the
one from 2.7 in regards of the two issues.

**Comment by mwperina** [comment]
This fix needs to be backported to Ansible 2.8.1, otherwise we couldn't easily use Ansible 2.8 in oVirt/RHV and we would need to distribute the module within all affected roles. It's already bad that this fix haven't been included in Ansible 2.8.0 (and that's another discussion why it happened), but if OpenStack module maintainers agrees with the backport, let's get it backported

**Comment by mwperina** [comment]
shipit

**Comment by emonty** [comment]
@abadger in this case I agree with @dominikholler - the patch should have been in 2.8 in the first place but we didn't get it landed in time.

The value we're passing as a default here - 'normal' - is the same as not passing the parameter, so for people who have clouds that support binding:vnic_type there will be no behavior change. HOWEVER - that parameter is handled by an API extension that may not be present on all OpenStack instalations - so without this patch there are some people who are broken.

**Comment by emonty** [comment]
shipit

**Comment by abadger** [comment]
<nod>  I don't think this is my call... The only decision making body we have is the core irc meeting.  Take it there so they can make a decision and then I can merge the backport if that's the decision.

File: lib/ansible/modules/cloud/openstack/os_port.py
Comment (abadger) [reply lib/ansible/modules/cloud/openstack/os_port.py:360]: Yes, that would be the way to do it.  Also, just for clarity, those are not blockers for merging this PR.  The question of behaviour change is the only thing holding this up.

**Comment by abadger** [comment]
A point in favour of merging this: the vnic_type parameter is only present in the 2.8.x branch.  So only 2.8.0 and 2.8.1 will have the parameter with the vnic_type default of normal by the time this is merged.

**Comment by Gestahlt** [comment]
Running Ansible 2.8.0
Our Openstack provider does not provide this parameter. 
I'm happy that you guys keep working on the OS modules. Sadly, it broke our instance creation plays.
As workaround i use now the naked API to create new NICs or modify them. 
Please fix this ASAP.

**Comment by abadger** [comment]
Approved at today's meeting.

**Comment by mwperina** [comment]
Thanks a lot Toshio!

@abadger

File: lib/ansible/modules/cloud/openstack/os_port.py
Comment (dominikholler) [reply lib/ansible/modules/cloud/openstack/os_port.py:360]: @abadger Please find the requested change in https://github.com/ansible/ansible/pull/57838

**Pull request merged**
